src/lib/profiles/data-management/Current/LoggingManagement.cpp

```diff
@@ -436,6 +437,210 @@ void LoggingManagement::DestroyLoggingManagement(void)
+    {
+        // Get the specified event buffer.
+        CircularEventBuffer * circular_event_buffer = GetImportanceBuffer((ImportanceType)importanceType);
+
+        TLVType container;
+        err = writer.StartContainer(ContextTag(importanceType), kTLVType_Structure, container);
+        SuccessOrExit(err);
+
+        err = writer.PutBytes(ContextTag(kTag_PersistEvent_EventBuffer), circular_event_buffer->mBuffer.GetQueue(), circular_event_buffer->mBuffer.GetQueueSize());
+        SuccessOrExit(err);
+
+        err = writer.Put(ContextTag(kTag_PersistEvent_EventBufferQueueHeadOffset), circular_event_buffer->mBuffer.QueueHead() - circular_event_buffer->mBuffer.GetQueue());
+        SuccessOrExit(err);
+
+        err = writer.Put(ContextTag(kTag_PersistEvent_EventBufferQueueLength), circular_event_buffer->mBuffer.DataLength());
+        SuccessOrExit(err);
+
+        err = writer.Put(ContextTag(kTag_PersistEvent_FirstEventId), circular_event_buffer->mFirstEventID);
+        SuccessOrExit(err);
+
+        err = writer.Put(ContextTag(kTag_PersistEvent_LastEventId), circular_event_buffer->mLastEventID);
+        SuccessOrExit(err);
+
+        err = writer.Put(ContextTag(kTag_PersistEvent_FirstEventTimestamp), circular_event_buffer->mFirstEventTimestamp);
+        SuccessOrExit(err);
+
+        err = writer.Put(ContextTag(kTag_PersistEvent_LastEventTimestamp), circular_event_buffer->mLastEventTimestamp);
+        SuccessOrExit(err);
+
+#if WEAVE_CONFIG_EVENT_LOGGING_UTC_TIMESTAMPS
+        err = writer.Put(ContextTag(kTag_PersistEvent_FirstEventUTCTimestamp), circular_event_buffer->mFirstEventUTCTimestamp);
+        SuccessOrExit(err);
+
+        err = writer.Put(ContextTag(kTag_PersistEvent_LastEventUTCTimestamp), circular_event_buffer->mLastEventUTCTimestamp);
+        SuccessOrExit(err);
+
+        err = writer.PutBoolean(ContextTag(kTag_PersistEvent_UTCInitialized), circular_event_buffer->mUTCInitialized);
+        SuccessOrExit(err);
+
+        err = writer.EndContainer(container);
+        SuccessOrExit(err);
+#endif
+    }
+
+    err = writer.EndContainer(rootContainer);
+    SuccessOrExit(err);
+    err = writer.Finalize();
+    SuccessOrExit(err);
+
+    serializedLen = (uint32_t)writer.GetLengthWritten();
+
+exit:
+    Platform::CriticalSectionExit();
+    return err;
+}
+
+/**
+ * @brief Load previously persisted Weave event.
+ *
+ */
+WEAVE_ERROR LoggingManagement::LoadEvent(uint8_t * serializedEvent, uint32_t serializedEventLen)
+{
+    WEAVE_ERROR err = WEAVE_NO_ERROR;
+
+    TLVReader reader;
+    reader.Init(serializedEvent, serializedEventLen);
+
+    TLVType rootContainer;
+    err = reader.Next(kTLVType_Structure, ProfileTag(kWeaveProfile_Common, kTag_SerializedEvent));
+    SuccessOrExit(err);
+    err = reader.EnterContainer(rootContainer);
+    SuccessOrExit(err);
+
+    for (int importanceType = kImportanceType_First; importanceType < kImportanceType_Last + 1; importanceType++)
```

Comment:
This will crash if kImportanceType_Last is greater than mMaxImportanceBuffer.

Reply:
Okay I looked at this code again... I'm not sure if my understanding is right - Seems like the number of buffers and their importance levels are hardcoded in nlError nl::Weave::Profiles::DataManagement::Platform::Init(). It not related to kImportanceType_Last, and not event mMaxImportanceBuffer? If this is true, should I pass the values as arguments and let platform decide how many buffers need to be persisted?

Reply:
No. The number of buffers is determined by the second argument to LoggingManagement() constructor.

Instead of using importance and calling GetImportanceBuffer(), you need to start with mEventBuffer and walk the chain of CircularEventBuffers via the mNext pointer until you reach the end (mNest == NULL).

Reply:
I see. Thanks.